### PR TITLE
New module: pnpm package manager

### PR DESCRIPTION
**Pull request by aretrosen**
##### SUMMARY

Support for the pnpm package manager

##### ISSUE TYPE

New Module/Plugin Pull Request

##### COMPONENT NAME

`pnpm`

##### ADDITIONAL INFORMATION

To the best of my knowledge, there currently isn't an Ansible module for pnpm available in the ansible-collection/community.general repository, unlike the existing modules for yarn and npm. pnpm is gaining popularity in the JavaScript ecosystem due to its speed and disk space efficiency, and many developers prefer it over npm and yarn. Given the strong resemblance between pnpm commands and npm commands, I have developed this module based on the npm module, resulting in significant similarities between the two codes.

Before proceeding with adding tests to this module, I would appreciate clarification on whether the pnpm module itself would be considered acceptable as a pull request.

### CODE REVIEW

File: plugins/modules/pnpm.py [hunk 1]
```diff
@@ -0,0 +1,412 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+
+# (c) 2023 Aritra Sen <aretrosen@proton.me>
```
Comment (felixfontein) [inline plugins/modules/pnpm.py:4]:
```suggestion
# Copyright (c) 2023 Aritra Sen <aretrosen@proton.me>
```

File: plugins/modules/pnpm.py [hunk 2]
```diff
@@ -0,0 +1,412 @@
...
+
+DOCUMENTATION = """
+---
+module: pnpm
+short_description: Manage node.js packages with pnpm
+description:
+  - Manage node.js packages with the pnpm package manager (https://pnpm.io/)
+author:
+  - "Aritra Sen (@aretrosen)"
+  - "Chris Hoffman (@chrishoffman), creator of NPM Ansible module)"
```
Comment (felixfontein) [inline plugins/modules/pnpm.py:23]:
```suggestion
  - "Chris Hoffman (@chrishoffman), creator of NPM Ansible module"
```

File: plugins/modules/pnpm.py [hunk 3]
```diff
@@ -0,0 +1,412 @@
...
+
+__metaclass__ = type
+
+
+DOCUMENTATION = """
+---
+module: pnpm
+short_description: Manage node.js packages with pnpm
+description:
+  - Manage node.js packages with the pnpm package manager (https://pnpm.io/)
```
Comment (felixfontein) [inline plugins/modules/pnpm.py:20]:
```suggestion
  - Manage node.js packages with the L(pnpm package manager, https://pnpm.io/).
```

File: plugins/modules/pnpm.py [hunk 4]
```diff
@@ -0,0 +1,412 @@
...
+  - community.general.attributes
+attributes:
+  check_mode:
+    support: full
+  diff_mode:
+    support: none
+options:
+  name:
+    description:
+      - The name of a node.js library to install
```
Comment (felixfontein) [inline plugins/modules/pnpm.py:34]: Please make sure that every description paragraph consists out of complete sentence(s).
```suggestion
      - The name of a node.js library to install.
```

File: plugins/modules/pnpm.py [hunk 5]
```diff
@@ -0,0 +1,412 @@
...
+  global:
+    description:
+      - Install the node.js library globally
+    required: false
+    default: false
+    type: bool
+  executable:
+    description:
+      - The executable location for pnpm.
+      - The default location it searches for is $PNPM_HOME, fails if not set.
```
Comment (felixfontein) [inline plugins/modules/pnpm.py:57]:
```suggestion
      - The default location it searches for is C($PNPM_HOME), fails if not set.
```

File: plugins/modules/pnpm.py [hunk 6]
```diff
@@ -0,0 +1,412 @@
...
+    type: path
+    required: false
+  ignore_scripts:
+    description:
+      - Use the --ignore-scripts flag when installing.
```
Comment (felixfontein) [inline plugins/modules/pnpm.py:62]:
```suggestion
      - Use the C(--ignore-scripts) flag when installing.
```
See https://docs.ansible.com/ansible/devel/dev_guide/developing_modules_documenting.html#format-macros-within-module-documentation and sections near it for more information on Ansible markup.

File: plugins/modules/pnpm.py [hunk 7]
```diff
@@ -0,0 +1,412 @@
...
+    required: false
+    type: bool
+    default: false
+  no_optional:
+    description:
+      - Don't install optional packages, equivalent to --no-optional.
```
Comment (felixfontein) [inline plugins/modules/pnpm.py:68]:
```suggestion
      - Do not install optional packages, equivalent to C(--no-optional).
```

**Review by felixfontein — COMMENTED** [review]
Thanks for your contribution! Please note that every new module must come with tests (integration and/or unit tests).

File: plugins/modules/pnpm.py [hunk 8]
```diff
@@ -0,0 +1,412 @@
...
+
+from __future__ import absolute_import, division, print_function
+
+__metaclass__ = type
+
+
+DOCUMENTATION = """
+---
+module: pnpm
+short_description: Manage node.js packages with pnpm
```
Comment (felixfontein) [inline plugins/modules/pnpm.py:18]:
```suggestion
short_description: Manage node.js packages with pnpm
version_added: 7.1.0
```

File: plugins/modules/pnpm.py [hunk 9]
```diff
@@ -0,0 +1,412 @@
...
+    type: path
+    required: false
+  version:
+    description:
+      - The version of the library to be installed, in semver format.
+    type: str
+    required: false
+  global:
+    description:
+      - Install the node.js library globally
```
Comment (felixfontein) [inline plugins/modules/pnpm.py:50]:
```suggestion
      - Install the node.js library globally.
```

File: plugins/modules/pnpm.py [hunk 10]
```diff
@@ -0,0 +1,412 @@
...
+  no_optional:
+    description:
+      - Do not install optional packages, equivalent to C(--no-optional).
+    required: false
+    type: bool
+    default: false
+  production:
+    description:
+      - Install dependencies in production mode.
+      - Pnpm will ignore any dependencies under devDependencies in package.json
```
Comment (felixfontein) [inline plugins/modules/pnpm.py:75]:
```suggestion
      - Pnpm will ignore any dependencies under devDependencies in package.json.
```

File: plugins/modules/pnpm.py [hunk 11]
```diff
@@ -0,0 +1,412 @@
...
+    required: false
+    type: bool
+    default: false
+  dev:
+    description:
+      - Install dependencies in development mode.
+      - Pnpm will ignore any regular dependencies in package.json
```
Comment (felixfontein) [inline plugins/modules/pnpm.py:82]:
```suggestion
      - Pnpm will ignore any regular dependencies in package.json.
```

File: plugins/modules/pnpm.py [hunk 12]
```diff
@@ -0,0 +1,412 @@
...
+  optional:
+    description:
+      - Install dependencies in optional mode.
+    required: false
+    default: false
+    type: bool
+  state:
+    description:
+      - Installation state of the named node.js library
+      - If absent is selected, a name option must be provided
```
Comment (felixfontein) [inline plugins/modules/pnpm.py:95]:
```suggestion
      - Installation state of the named node.js library.
      - If absent is selected, a name option must be provided.
```

File: plugins/modules/pnpm.py [hunk 13]
```diff
@@ -0,0 +1,412 @@
...
+    type: str
+    required: false
+    default: present
+    choices: [ "present", "absent", "latest" ]
+requirements:
+    - Pnpm executable present in $PATH, preferably as $PNPM_HOME.
```
Comment (felixfontein) [inline plugins/modules/pnpm.py:101]:
```suggestion
    - Pnpm executable present in C($PATH), preferably as C($PNPM_HOME).
```

File: plugins/modules/pnpm.py [hunk 14]
```diff
@@ -0,0 +1,412 @@
...
+  global:
+    description:
+      - Install the node.js library globally
+    required: false
+    default: false
+    type: bool
+  executable:
+    description:
+      - The executable location for pnpm.
+      - The default location it searches for is C($PNPM_HOME), fails if not set.
```
Comment (felixfontein) [inline plugins/modules/pnpm.py:57]: `PNPM_HOME` is not mentioned in the module's code.

**Comment by felixfontein** [comment]
Please do not forget about the tests, and please remove the `@` from that commit. Thanks.

File: plugins/modules/pnpm.py [hunk 15]
```diff
@@ -0,0 +1,448 @@
...
+  diff_mode:
+    support: none
+options:
+  name:
+    description:
+      - The name of a node.js library to install.
+      - All packages in package.json are installed if not provided.
+    type: str
+    required: false
+   alias:
```
Comment (felixfontein) [inline plugins/modules/pnpm.py:39]: Here's one space too much:
```suggestion
  alias:
```

File: tests/unit/plugins/modules/test_npm.py [hunk 16]
```diff
@@ -211,7 +211,7 @@ def test_present_package_json_production(self):
             'production': 'true',
             'global': 'true',
             'state': 'present',
-        })
+
```
Comment (felixfontein) [inline tests/unit/plugins/modules/test_npm.py:214]: This change doesn't work.

**Comment by aretrosen** [comment]
This pr is closed in favor of a new PR, as I made some wrong changes in this pr.

**Pull request closed without merging**